includes/oauth/class-oauth.php

```diff
@@ -25,6 +25,17 @@ public static function get_proxy_api_key() {
 		return get_option( NEWSPACK_MANAGER_API_KEY_OPTION_NAME );
 	}
 
+	/**
+	 * Get unique ID for user or client.
+	 */
+	private static function get_unique_id() {
+		$id = Reader_Activation::get_client_id();
+		if ( ! $id ) {
+			$id = get_current_user_id();
+		}
+		return $id;
+	}
```

Comment:
This method is not ensuring that the reader receives a unique ID and there's a security concern.

`$id` should only be the current user's ID if we can verify that the user is logged in, and IMO it should be prioritized over any other method.

`Reader_Activation::get_client_id();` is user input. Since this is performed during the request (uncached) it is safer to have this method generate a random value, set it as cookie, and later check the validity.

Reply:
> `Reader_Activation::get_client_id();` is user input. Since this is performed during the request (uncached) it is safer to have this method generate a random value, set it as cookie, and later check the validity.

Does this value even need to be directly associated with the reader's identity? Can't it just be a random, unique string that can be verified within the same session, in which case we could use a `nonce`?

Reply:
It's my understanding that it's what it does. A custom CSRF protection implementation uses this unique value to validate the request after it returns from the proxy/third party.

We must be careful on generating those, though. Including nonces. The initial state is anonymous and subjected to cache. This random value must be generated on the request and preferably by the backend to ensure uniqueness, otherwise we may get duplicate session tokens.

Reply:
(nonces and caching discussion [has started here](https://github.com/Automattic/newspack-plugin/pull/1780#discussion_r922332663))

Reply:
Although the mentioned discussion touch similar topics, the `nonce` used at `/wp-json/newspack/v1/login/google/register` the `$id` used at `Reader_Activation::get_client_id();` have different goals, right?

Reply:
1. _Nonces vs. caching_: removed nonce use in 9965c61c98cecd649d5d366a8c99f126b130fe43. I was originally introduced to authenticate the registration request, but after all the changes in the meantime, this seems not necessary anymore. The request to `/login/google/register` is fine without the nonce. 
2. _Unique id for 3rd party data transient_: I've changed it so it tries for user ID, then then the client ID cookie, and then as a last resort, the session id. [VIP-Coding-Standards discourages the use of `session_id`](https://github.com/Automattic/VIP-Coding-Standards/blob/e1eb7c9dbf93ebcbb20a610c6babe4b74725551d/WordPressVIPMinimum/Sniffs/Functions/RestrictedFunctionsSniff.php#L168), but I'm not sure why. Session ID seems like an idea candidate here.